llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```
@@ -905,6 +903,14 @@ bool SPIRVInstructionSelector::selectExtInst(Register ResVReg,
                                              const SPIRVType *ResType,
                                              MachineInstr &I,
                                              GL::GLSLExtInst GLInst) const {
+  if (!STI.canUseExtInstSet(
+          SPIRV::InstructionSet::InstructionSet::GLSL_std_450)) {
+    std::string DiagMsg;
+    raw_string_ostream OS(DiagMsg);
+    I.print(OS, true, false, false, false);
+    DiagMsg += " is only supported with the GLSL extended instruction set.\n";
+    report_fatal_error(DiagMsg.c_str(), false);
+  }
```

IMO this isn't the right way to do this kind of error handling. `report_fatal_error` just causes the compiler to crash, but given that this can happen given IR for the wrong target just crashing because of it isn't great. We should probably be using `LLVMContext::diagnose` here instead so that we can bubble the error up to the driver.  Note that I do think we probably need to do a little bit of refactoring to use `diagnose` here, as this will probably just crash anyway with a `Cannot select` error if we just call diagnose and return from this function.

I'll comment on #123847 to this effect.

In the meantime, I also don't really like that we're adding this error handling only for the `GLSLExtInst` overload and not all of the extremely related cases. I think we should treat this as out of scope for the `reflect` function change, remove the `report_fatal_error`, and drop or XFAIL the `reflect-error.ll` test for now - we can reintroduce that test when we fix #123847.

---

I'm not sure I agree. The other intrinsic selections in  `selectIntrinsic` are doing `report_fatal_error` The only difference is I don't see the report fatal error code pathes are being tested. Switchinng to `LLVMContext::diagnose` would be a new pattern and if it requires a refactor like you claim we probably want to involve the SPIRV backend stake holders depending on how large of a refactor you imagine.

Second this is a case that should never be possible. The code that emits the reflect intrinsic is only emited when the target is `spirv` via the spirv target builtins. However the opencl test is  using the `spirv32` and `spirv64` targets. This intrinsic should be invalid if the target isn't `spirv` and to me it makes sense to fail immediately.

Further their is no mixed mode where you can do ExtInst in opencl and opengl so checking `canUseExtInstSet` seems like the right way forward.

Third I agree that the error handling in `GLSLExtInst` should have its reciprocal error in the `OpenCLExtInst`. However thats beyond the scope of this PR. I wouldn't want to add error handling like that without a test and I think a test like that is out of scope for a change about the `reflect` api

---

> I'm not sure I agree. The other intrinsic selections in `selectIntrinsic` are doing `report_fatal_error` The only difference is I don't see the report fatal error code pathes are being tested. Switching to `LLVMContext::diagnose` would be a new pattern and if it requires a refactor like you claim we probably want to involve the SPIRV backend stake holders depending on how large of a refactor you imagine.

I am indeed asserting that many of the other uses of report_fatal_error in this file are also wrong. However, you're probably right that it makes sense to the existing pattern for now and clean this all up in a cohesive change.

> Second this is a case that should never be possible. The code that emits the reflect intrinsic is only emited when the target is `spirv` via the spirv target builtins. However the opencl test is using the `spirv32` and `spirv64` targets. This intrinsic should be invalid if the target isn't `spirv` and to me it makes sense to fail immediately.

If the case is impossible why isn't this just an assert? In any case "failing immediately" and "causing the program to crash" are different things, and `report_fatal_error` is the latter.

> Further their is no mixed mode where you can do ExtInst in opencl and opengl so checking `canUseExtInstSet` seems like the right way forward.
> 
> Third I agree that the error handling in `GLSLExtInst` should have its reciprocal error in the `OpenCLExtInst`. However thats beyond the scope of this PR. I wouldn't want to add error handling like that without a test and I think a test like that is out of scope for a change about the `reflect` api

IMO this is an argument for making a change that adds error handling to both functions *before* this PR, so that we don't end up in a state where the codebase is inconsistent.